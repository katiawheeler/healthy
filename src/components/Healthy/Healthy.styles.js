import styled from '@emotion/styled';

export const BannerWrapper = styled('div')`
  color: white;
  background: #ef5350;
  width: 100%;
  position: relative;
`;

export const BannerContent = styled('div')`
  padding: 20px;
  max-width: 1000px;
  font-size: 16px;
  font-family: 'Helvetica', 'Arial', sans-serif;
  margin: 0 auto;
  text-align: center;
`;

export const CloseButton = styled('button')`
  position: absolute;
  top: 30%;
  right: 30px;
  background: transparent;
  color: white;
  border: none;
  box-shadow: none;
  cursor: pointer;
  font-size: 16px;
  font-family: 'Helvetica','Arial',sans-serif;

  &:hover {
    background: #e53935;
  }

  &:active,
  &:focus {
    outline: none;
  }
`;
