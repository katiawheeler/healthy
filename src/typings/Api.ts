export interface Api {
  /** Name of the API - to be displayed in the banner **/
  name: string;
  /** Endpoint of the API to ping **/
  endpoint: string;
  /** Custom message to be displayed when there's an error **/
  message?: string;
}

export interface Response {
  /** HTTP Error Code **/
  code: number;
  /** Status message **/
  message: string;
}

export interface ApiWithResponse {
  api: Api;
  response: Response;
}
