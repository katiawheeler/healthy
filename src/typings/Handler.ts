import { Api, Response } from './Api';

export type Handler = (api: Api, response?: Response) => void;
