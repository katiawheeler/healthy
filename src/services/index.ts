import { Api, Response as ApiResponse } from '../typings/Api';
import { Handler } from '../typings/Handler';
 
/** Begins the chain of API calls */
export const begin = async (apis: Api[], onSuccess: Handler, onError:Handler, interval?: number) => {
 const currentInterval = interval ? interval : 30000;
 
 apis.forEach(async a => {
   await makeCall(a, onSuccess, onError);
 });
 
 /** Set the timeout and do it again */
 setTimeout(async () => await begin(apis, onSuccess, onError, currentInterval), currentInterval);
};
 
/** Make the actual fetch and return a Response object */
const makeCall = async (api: Api, onSuccess: Handler, onError:Handler) => {


 await fetch(api.endpoint)
   .then(apiResponse => handleResponse(api, onSuccess, onError, apiResponse))
   // tslint:disable-next-line
   .catch(error => handleResponse(api, onSuccess, onError));
};
 
/**
* Handles the Response
*/
const handleResponse = (api: Api, onSuccess: Handler, onError:Handler, response?: Response) => {

  if(response){

    const apiResponse: ApiResponse = {
      code: response.status,
      message: response.statusText,
    };

    (200 == apiResponse.code || 
      201 == apiResponse.code || 
      202 == apiResponse.code || 
      203 == apiResponse.code || 
      204 == apiResponse.code || 
      205 == apiResponse.code || 
      206 == apiResponse.code || 
      207 == apiResponse.code || 
      208 == apiResponse.code || 
      226 == apiResponse.code)  ? onSuccess(api, apiResponse) : onError(api, apiResponse);

  } else {
    onError(api);
  }

};

 
export default { begin, makeCall, handleResponse };