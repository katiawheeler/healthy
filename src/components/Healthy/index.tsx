import React from 'react';
import { begin } from '../../services';
import { Api, ApiWithResponse, Response } from '../../typings/Api';
import { Handler } from '../../typings/Handler';
import { BannerContent, BannerWrapper, CloseButton } from './Healthy.styles';

export interface HealthyProps {
  /** An array of API objects */
  apis: Api[];
  /** A callback that's called after the component handles the error, for additional error handling */
  onError?: Handler;
  /** The interval at which to call the APIs in milliseconds; default is 30 seconds (30000) */
  interval?: number;
  /** A callback to handle all responses that are not errors, in case
   * you want to handle different responses differently
   */
  onResponse?: Handler;
  /** CSS class names to assign to the banner, banner content, and close button */
  classes?: {
    banner?: string;
    content?: string;
    closeButton?: string;
  };
  /** Whether or not to show a close icon - default is false */
  closeable?: boolean;
}

export interface HealthyState {
  hasError: boolean;
  problemChildren: ApiWithResponse[];
}

class Healthy extends React.Component<HealthyProps> {
  public state: HealthyState = {
    hasError: false,
    problemChildren: [],
  };

  public componentDidMount = async () => {
    await begin(this.props.apis, this.handleError, this.props.interval, this.props.onResponse || undefined);
  };

  public handleError = (api: Api, response: Response) => {
    const problemChildren = this.state.problemChildren;
    if (problemChildren.find(item => item.api.endpoint === api.endpoint) === undefined) {
      problemChildren.push({ api, response });
    }
    this.setState({
      problemChildren,
      hasError: true,
    });
    if (this.props.onError) { this.props.onError(api, response); }
  };

  public handleClose = () => this.setState({ hasError: false });

  public determineMessage = () => {
    if (this.state.problemChildren.length === 0) { return; }
    const firstApi: ApiWithResponse = this.state.problemChildren[0];

    // set to default message or message provided
    let message = firstApi.api.message ? firstApi.api.message : `We are currently experiencing issues with our ${
      firstApi.api.name
    } service`;

    // if there are more than one APIs down, show how many services are down
    if (this.state.problemChildren.length > 1) {
      message = `We are currently experiencing issues with ${this.state.problemChildren.length} services`;
    }

    return message;
  };

  public render() {
    const message = this.determineMessage();
    return (
      this.state.hasError && (
        <BannerWrapper className={this.props.classes && this.props.classes.banner}>
          <BannerContent className={this.props.classes && this.props.classes.content}>
            <span>{message}</span>
          </BannerContent>
          {this.props.closeable && (
            <CloseButton
              onClick={this.handleClose}
              className={this.props.classes && this.props.classes.closeButton}
            >
              x
            </CloseButton>
          )}
        </BannerWrapper>
      )
    );
  }
}

export default Healthy;
