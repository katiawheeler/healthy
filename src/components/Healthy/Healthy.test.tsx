import { shallow } from 'enzyme';
import React from 'react';
import Healthy, { HealthyProps } from '.';
import { begin } from '../../services/';
import { Api, Response } from '../../typings/Api';

jest.mock('../../services/');

const goodEndpoint = 'https://httpstat.us/200';
const badEndpoint = 'https://httpstat.us/404';

describe('src/components/Healthy', () => {
  describe('componentDidMount', () => {
    it('should call begin', () => {
      const props: HealthyProps = {
        apis: [
          {
            endpoint: goodEndpoint,
            name: 'Test Api',
          },
        ],
      };

      shallow(<Healthy {...props} />);
      expect(begin).toHaveBeenCalledTimes(1);
    });
  });

  describe('handleError', () => {
    const api: Api = {
      endpoint: badEndpoint,
      name: 'Bad Api',
    };
    const response: Response = {
      code: 404,
      message: 'Not found',
    };

    const props: HealthyProps = {
      apis: [api],
      onError: jest.fn(),
    };

    describe('when the problemChild is not currently in state', () => {
      it('should add the faulty api into problemChildren', async () => {
        const component = shallow<Healthy>(<Healthy {...props} />);
        expect(component.state('problemChildren')).toEqual([]);
        await component.instance().handleError(api, response);
        expect(component.state('problemChildren')).toEqual([{ api, response }]);
      });
    });

    describe('when the problemChild is already in state', () => {
      it('should not duplicate the api in problemChildren', async () => {
        const component = shallow<Healthy>(<Healthy {...props} />);
        component.setState({ problemChildren: [{ api, response }] });
        await component.instance().handleError(api, response);
        expect(component.state('problemChildren')).toEqual([{ api, response }]);
      });
    });

    describe('when props.onError exists', () => {
      it('should call props.onError', async () => {
        const component = shallow<Healthy>(<Healthy {...props} />);
        await component.instance().handleError(api, response);
        expect(props.onError).toHaveBeenCalled();
      });
    });
  });

  describe('handleClose', () => {
    it('should set hasError to false on state', () => {
      const props: HealthyProps = {
        apis: [
          {
            endpoint: goodEndpoint,
            name: 'Test Api',
          },
        ],
      };
      const component = shallow<Healthy>(<Healthy {...props} />);

      component.setState({ hasError: true });
      expect(component.state('hasError')).toBeTruthy();
      component.instance().handleClose();
      expect(component.state('hasError')).toBeFalsy();
    });
  });

  describe('determineMessage', () => {
    describe('when problemChildren has a length === 0', () => {
      it('should return undefined', () => {
        const props: HealthyProps = {
          apis: [
            {
              endpoint: goodEndpoint,
              name: 'Test Api',
            },
          ],
        };
        const component = shallow<Healthy>(<Healthy {...props} />);
        expect(component.state('problemChildren')).toHaveLength(0);
        const result = component.instance().determineMessage();
        expect(result).toBe(undefined);
      });
    });

    describe('when problemChildren has a length === 1', () => {
      describe('when a custom message was passed', () => {
        it('should return the custom message', () => {
          const api: Api = {
            endpoint: badEndpoint,
            name: 'Bad Api',
            message: 'Hello, world',
          };
          const response: Response = {
            code: 404,
            message: 'Not found',
          };

          const props: HealthyProps = {
            apis: [api],
            onError: jest.fn(),
          };

          const component = shallow<Healthy>(<Healthy {...props} />);
          component.setState({
            problemChildren: [
              {
                api,
                response,
              },
            ],
          });
          expect(component.state('problemChildren')).toHaveLength(1);
          const result = component.instance().determineMessage();
          expect(result).toBe('Hello, world');
        });
      });

      describe('when no custom message was passed', () => {
        it('should return the default message', () => {
          const api: Api = {
            endpoint: badEndpoint,
            name: 'Bad Api',
          };
          const response: Response = {
            code: 404,
            message: 'Not found',
          };

          const props: HealthyProps = {
            apis: [api],
            onError: jest.fn(),
          };

          const component = shallow<Healthy>(<Healthy {...props} />);
          component.setState({
            problemChildren: [
              {
                api,
                response,
              },
            ],
          });
          expect(component.state('problemChildren')).toHaveLength(1);
          const result = component.instance().determineMessage();
          expect(result).toBe('We are currently experiencing issues with our Bad Api service');
        });
      });
    });

    describe('when problemChildren has a length > 1', () => {
      it('should list the number of services down', () => {
        const api: Api = {
          endpoint: badEndpoint,
          name: 'Bad Api',
        };
        const secondApi: Api = {
          endpoint: badEndpoint,
          name: 'Bad Api',
        };
        const response: Response = {
          code: 404,
          message: 'Not found',
        };

        const props: HealthyProps = {
          apis: [api, secondApi],
          onError: jest.fn(),
        };

        const component = shallow<Healthy>(<Healthy {...props} />);
        component.setState({
          problemChildren: [{ api, response }, { api: secondApi, response }],
        });
        expect(component.state('problemChildren')).toHaveLength(2);
        const result = component.instance().determineMessage();
        expect(result).toBe('We are currently experiencing issues with 2 services');
      });
    });
  });

  describe('render', () => {
    describe('when there are no problemChildren', () => {
      it('should match the snapshot', () => {
        const props: HealthyProps = {
          apis: [
            {
              endpoint: goodEndpoint,
              name: 'Api',
            },
          ],
        };
        const component = shallow(<Healthy {...props} />);
        expect(component).toMatchSnapshot();
      });
    });

    describe('when there is one problemChild', () => {
      describe('when no custom message is passed', () => {
        it('should match the snapshot', () => {
          const api: Api = {
            endpoint: badEndpoint,            
            name: 'Bad Api',
          };
          const response: Response = {
            code: 404,
            message: 'Not found',
          };
          const props: HealthyProps = {
            apis: [api],
          };
          const component = shallow<Healthy>(<Healthy {...props} />);
          component.setState({ problemChildren: [{ api, response }] });
          expect(component).toMatchSnapshot();
        });
      });
      describe('when there is a custom message is passed', () => {
        it('should match the snapshot', () => {
          const api: Api = {
            endpoint: badEndpoint,
            name: 'Bad Api',
            message: 'Custom Message, here',
          };
          const response: Response = {
            code: 404,
            message: 'Not found',
          };
          const props: HealthyProps = {
            apis: [api],
          };
          const component = shallow<Healthy>(<Healthy {...props} />);
          component.setState({ problemChildren: [{ api, response }] });
          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('when there are multiple problemChildren', () => {
      it('should match the snapshot', () => {
        const api: Api = {
          endpoint: badEndpoint,          
          name: 'Bad Api',
        };
        const secondApi: Api = {
          endpoint: badEndpoint,
          name: 'Bad Api',
        };
        const thirdApi: Api = {
          endpoint: badEndpoint,
          name: 'Bad Api',
        };
        const response: Response = {
          code: 404,
          message: 'Not found',
        };
        const props: HealthyProps = {
          apis: [api, secondApi, thirdApi],
        };
        const component = shallow<Healthy>(<Healthy {...props} />);
        component.setState({
          problemChildren: [
            { api, response },
            { api: secondApi, response },
            { api: thirdApi, response },
          ],
        });
        expect(component).toMatchSnapshot();
      });
    });
  });
});
