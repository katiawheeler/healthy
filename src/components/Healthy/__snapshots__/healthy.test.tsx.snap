// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/Healthy render when there are multiple problemChildren should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Healthy
    apis={
      Array [
        Object {
          "endpoint": "https://httpstat.us/404",
          "name": "Bad Api",
        },
        Object {
          "endpoint": "https://httpstat.us/404",
          "name": "Bad Api",
        },
        Object {
          "endpoint": "https://httpstat.us/404",
          "name": "Bad Api",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): false,
  Symbol(enzyme.__nodes__): Array [
    false,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`src/components/Healthy render when there are no problemChildren should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Healthy
    apis={
      Array [
        Object {
          "endpoint": "https://httpstat.us/200",
          "name": "Api",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): false,
  Symbol(enzyme.__nodes__): Array [
    false,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`src/components/Healthy render when there is one problemChild when no custom message is passed should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Healthy
    apis={
      Array [
        Object {
          "endpoint": "https://httpstat.us/404",
          "name": "Bad Api",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): false,
  Symbol(enzyme.__nodes__): Array [
    false,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`src/components/Healthy render when there is one problemChild when there is a custom message is passed should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Healthy
    apis={
      Array [
        Object {
          "endpoint": "https://httpstat.us/404",
          "message": "Custom Message, here",
          "name": "Bad Api",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): false,
  Symbol(enzyme.__nodes__): Array [
    false,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
